### PR TITLE
[LV][EVL]Explicitly model AVL as sub, original TC, EVL_PHI.

### PR DESCRIPTION
**Pull request by alexey-bataev**
Patch explicitly models AVL as sub original TC, EVL_PHI instead of
having it in EXPLICIT-VECTOR-LENGTH VPInstruction. Required for correct
safe dependence distance suport.

### CODE REVIEW

**Comment by alexey-bataev** [comment]
Ping!

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp [hunk 1]
```diff
@@ -1417,9 +1418,13 @@ bool VPlanTransforms::tryAddExplicitVectorLength(VPlan &Plan) {
   // Create the ExplicitVectorLengthPhi recipe in the main loop.
   auto *EVLPhi = new VPEVLBasedIVPHIRecipe(StartV, DebugLoc());
   EVLPhi->insertAfter(CanonicalIVPHI);
-  auto *VPEVL = new VPInstruction(VPInstruction::ExplicitVectorLength,
-                                  {EVLPhi, Plan.getTripCount()});
-  VPEVL->insertBefore(*Header, Header->getFirstNonPhi());
+  // Compute original TC - IV as the AVL (requested vector length).
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:1421]:
```suggestion
  // Compute original TC - IV as the AVL (application vector length).
```

Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:1421]: nit: the TODO to handle MaxSafeDist should move from GetEVL above to here where the AVL argument is prepared for EVL.

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 2]
```diff
@@ -484,10 +484,7 @@ Value *VPInstruction::generatePerPart(VPTransformState &State, unsigned Part) {
     // TODO: Restructure this code with an explicit remainder loop, vsetvli can
     // be outside of the main loop.
     assert(Part == 0 && "No unrolling expected for predicated vectorization.");
-    // Compute VTC - IV as the AVL (requested vector length).
-    Value *Index = State.get(getOperand(0), VPIteration(0, 0));
-    Value *TripCount = State.get(getOperand(1), VPIteration(0, 0));
-    Value *AVL = State.Builder.CreateSub(TripCount, Index);
+    Value *AVL = State.get(getOperand(0), VPIteration(0, 0));
     Value *EVL = GetEVL(State, AVL);
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:488]: nit: simpler to inline GetEVL?

A follow-up thought: have a recipe for calling any intrinsic.

**Review by ayalz — APPROVED** [review]
LGTM, adding some minor nits.

**Pull request merged**